scipy/stats/stats.py

```diff
@@ -3294,26 +3294,31 @@ def spearmanr(a, b=None, axis=0, nan_policy='propagate'):
     """
     a, axisout = _chk_asarray(a, axis)
 
-    contains_nan, nan_policy = _contains_nan(a, nan_policy)
+    a_contains_nan, nan_policy = _contains_nan(a, nan_policy)
 
-    if contains_nan and nan_policy == 'omit':
+    if a_contains_nan and nan_policy == 'omit':
         a = ma.masked_invalid(a)
-        b = ma.masked_invalid(b)
```

Comment:
So it's a return of revenge of https://github.com/scipy/scipy/issues/6134, i.e. whether a and b represent paired values of not. A quick check with R:

```
> cor.test(c(1, 2, 3, 4), c(8, 7, 6, NA), method='spearman')
	Spearman's rank correlation rho

data:  c(1, 2, 3, 4) and c(8, 7, 6, NA)
S = 8, p-value = 0.3333
alternative hypothesis: true rho is not equal to 0
sample estimates:
rho 
 -1 

> cor.test(c(1, 2, 3), c(8, 7, 6), method='spearman')

	Spearman's rank correlation rho

data:  c(1, 2, 3) and c(8, 7, 6)
S = 8, p-value = 0.3333
alternative hypothesis: true rho is not equal to 0
sample estimates:
rho 
 -1 
```

I think it's a good idea to match R's behavior, and have a test with a number-to-number comparison against R. See e.g. `test_kendalltau` in `stats/tests/test_stats.py`

@vigna would you want to offer an opinion here?